### PR TITLE
some UTF-8 improvements

### PR DESCRIPTION
**Pull request by Goober5000**
1) Write all non-UTF-8 strings to JSON, either by converting or truncating
2) Add `coerce_to_utf8` function
3) Tweak `read_raw_file_text` because the function returns length on success, not 0

Fixes Mantis ticket:
http://scp.indiegames.us/mantis/view.php?id=3209

### CODE REVIEW

File: code/parse/parselo.cpp [hunk 1]
```diff
@@ -2258,6 +2260,74 @@ void read_raw_file_text(const char *filename, int mode, char *raw_text)
...
+		return;
+	}
+
+	bool isLatin1 = util::guessLatin1Encoding(str, len);
+
+	// we can convert it
+	if (isLatin1)
+	{
+		size_t newlen = len;
+		auto newstr = new char[newlen];
```
Comment (asarium) [inline code/parse/parselo.cpp:2283]: Please use `std::unique_ptr<char[]>` for a bit more safety.

File: code/pilotfile/JSONFileHandler.cpp [hunk 2]
```diff
@@ -4,6 +4,11 @@
 #include "libs/jansson.h"
 #include "parse/parselo.h"
 
+// it's kinda tricky to #include "utf.h" here, so we forward-declare this, but note that it's a C function
+extern "C" {
+	extern int utf8_check_string(const char *string, size_t length);
```
Comment (asarium) [inline code/pilotfile/JSONFileHandler.cpp:9]: What are the issues with including that header? We should avoid doing library declarations on our own if at all possible. The contents of this function also seem to be more or less the same as `utf8::find_invalid` so if including the header doesn't work maybe you could use that function instead?

Comment (Goober5000) [reply code/pilotfile/JSONFileHandler.cpp:9]: Just that it's a very convoluted path.  But I'll try that alternative.

File: code/parse/parselo.cpp (hunk 1 above)
Comment (Goober5000) [reply code/parse/parselo.cpp:2283]: Ok.

**Pull request merged**